ensime_shared/protocol_handler.py

```diff
@@ -0,0 +1,249 @@
+        for member in payload["members"]:
+            add(member, 1)
+
+    def handle_symbol_search(self, call_id, payload):
+        """Handler for symbol search results"""
+        self.log(payload)
+        syms = payload["syms"]
+        qfList = []
+        for sym in syms:
+            p = sym.get("pos")
+            if p:
+                item = self.to_quickfix_item(str(p["file"]),
+                                            p["line"],
+                                            str(sym["name"]),
+                                            "info")
+                qfList.append(item)
+        self.write_quickfix_list(qfList)
+
+    def handle_symbol_info(self, call_id, payload):
+        """Handler for response `SymbolInfo`."""
+        warn = lambda e: self.message("unknown_symbol")
+        with catch(KeyError, warn):
+            decl_pos = payload["declPos"]
+            f = decl_pos.get("file")
+            self.log(str(self.call_options[call_id]))
+            display = self.call_options[call_id].get("display")
+            if display and f:
+                self.vim.command(commands["display_message"].format(f))
+
+            open_definition = self.call_options[call_id].get("open_definition")
+            if open_definition and f:
+                self.clean_errors()
+                self.vim_command("doautocmd_bufleave")
+                split = self.call_options[call_id].get("split")
+                vert = self.call_options[call_id].get("vert")
+                key = ""
+                if split:
+                    key = "vert_split_window" if vert else "split_window"
+                else:
+                    key = "edit_file"
+                self.vim.command(commands[key].format(f))
+                self.vim_command("doautocmd_bufreadenter")
+                self.set_position(decl_pos)
+                del self.call_options[call_id]
+
+    def handle_string_response(self, call_id, payload):
+        """Handler for response `StringResponse`.
+
+        This is the response for the following requests:
+          1. `DocUriAtPointReq` or `DocUriForSymbolReq`
+          2. `DebugToStringReq`
+          3. `FormatOneSourceReq`
+        """
+        self.log(str(payload))
+        self.handle_doc_uri(call_id, payload)
+
+    def handle_doc_uri(self, call_id, payload):
+        """Handler for responses of Doc URIs."""
+        if not self.en_format_source_id:
+            self.log("handle_string_response: received doc path")
+            port = self.ensime.http_port()
+
+            url = payload["text"]
+
+            if not url.startswith("http"):
+                url = gconfig["localhost"].format(port, payload["text"])
+
+            browse_enabled = self.call_options[call_id].get("browse")
+
+            if browse_enabled:
+                log_msg = "handle_string_response: browsing doc path {}"
+                self.log(log_msg.format(url))
+                try:
```

Comment:
Sorry, `catch` calls `str()` on the exception rather than getting it's message attribute, I was mistaken. I had to re-read the implementation.